### PR TITLE
more correct pt definition in writing out ntuples

### PR DESCRIPTION
**Pull request by YonsiG**
The original writing out ntuples is not a good estimation of the object pt, and we would prefer to use the pLS pt for pT3 and pT5 objects. Use pt from innerRadius to calculate the T5 pt in the output definition, instead of calculating from inner T3 and outer T3 "beta"s. This slide shows that, in the reco tracks that are matched to true track, the new pt have a smaller difference to the sim_pt compared to master.
On the other hand, there is still room to improve. We see that in the pT5, average over pLS pt and T5 inner radius pt has better resolution than only using pLS pt. We want to also test this on pT3. Then we realize that the radius  of T3s aer not defined in the algorithm until we build a T5, which is not optimal and cannot be taken out as the T3 pt. This would come in a later PR to move the T3 radius calculation earlier, so that the output pt of pT3 can change accordingly.

### CODE REVIEW

**Comment by YonsiG** [comment]
Related slides showing the summary of this study
https://indico.cern.ch/event/1374270/contributions/5775805/attachments/2836918/4959541/output%20pt%20definition%20summary%20details.pdf

File: code/core/write_sdl_ntuple.cc [hunk 1]
```diff
@@ -999,6 +962,7 @@ std::tuple<float, float, float, vector<unsigned int>, vector<unsigned int>> pars
 std::tuple<float, float, float, vector<unsigned int>, vector<unsigned int>> parseT5(SDL::Event<SDL::Acc>* event, unsigned int idx)
 {
     SDL::trackCandidatesBuffer<alpaka::DevCpu>& trackCandidatesInGPU = (*event->getTrackCandidates());
+    SDL::quintupletsBuffer<alpaka::DevCpu>& quintupletsInGPU = (*event->getQuintuplets());
     SDL::tripletsBuffer<alpaka::DevCpu>& tripletsInGPU = (*event->getTriplets());
```
Comment (VourMa) [inline code/core/write_sdl_ntuple.cc:966]: This is not needed anymore.

File: code/core/write_sdl_ntuple.cc
Comment (VourMa) [inline code/core/write_sdl_ntuple.cc:1]: Could you remove the lines that set `kRinv1GeVf` and use
https://github.com/SegmentLinking/TrackLooper/blob/a8bb417aac4b71d703a1e3c9f62bf02aeeff6941/SDL/Constants.h#L158
instead? It's redefined in almost all of the functions, and this is far from ideal.

Similarly for `k2Rinv1GeVf`.

Also, if any of those is not used, please just remove it.

**Review by VourMa — COMMENTED** [review]
Thank you for this PR, @YonsiG. The changes seem good. I have left some proposals for further clean-up relevant to your changes.

File: code/core/write_sdl_ntuple.cc [hunk 2]
```diff
@@ -814,6 +805,7 @@ std::tuple<float, float, float, vector<unsigned int>, vector<unsigned int>> pars
 {
     // Get relevant information
     SDL::trackCandidatesBuffer<alpaka::DevCpu>& trackCandidatesInGPU = (*event->getTrackCandidates());
+    SDL::quintupletsBuffer<alpaka::DevCpu>& quintupletsInGPU = (*event->getQuintuplets());
     SDL::tripletsBuffer<alpaka::DevCpu>& tripletsInGPU = (*event->getTriplets());
```
[the remark this answers is not in the document]
Comment (YonsiG) [reply code/core/write_sdl_ntuple.cc:809]: It is still needed in getting T5 inner radius. pt of pT5 is then defined as (pt_pLS+pt_T5)/2

File: code/core/write_sdl_ntuple.cc [hunk 3]
```diff
@@ -829,15 +821,9 @@ std::tuple<float, float, float, vector<unsigned int>, vector<unsigned int>> pars
     //                            oo -- oo -- oo   second T3 of the T5
     unsigned int pT5 = trackCandidatesInGPU.directObjectIndices[idx];
     std::vector<unsigned int> Hits = getOuterTrackerHitsFrompT5(event, pT5);
```
[the remark this answers is not in the document]
Comment (YonsiG) [reply code/core/write_sdl_ntuple.cc:823]: Deleted. Thank you!

File: code/core/write_sdl_ntuple.cc [hunk 4]
```diff
@@ -964,28 +937,18 @@ std::tuple<float, float, float, vector<unsigned int>, vector<unsigned int>> pars
     // pLS            01    23    45               (anchor hit of a minidoublet is always the first of the pair)
```
[the remark this answers is not in the document]
Comment (YonsiG) [reply code/core/write_sdl_ntuple.cc:937]: Deleted

File: code/core/write_sdl_ntuple.cc (hunk 1 above)
Comment (YonsiG) [reply code/core/write_sdl_ntuple.cc:966]: Deleted

File: code/core/write_sdl_ntuple.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (VourMa) [reply code/core/write_sdl_ntuple.cc:809]: `float pt_T5 = __H2F(quintupletsInGPU.innerRadius[T5Index]) * kRinv1GeVf;`
I fail to see how `tripletsInGPU` is needed.

File: code/core/write_sdl_ntuple.cc
Comment (YonsiG) [reply code/core/write_sdl_ntuple.cc:1]: Updated, thanks!

File: code/core/write_sdl_ntuple.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (YonsiG) [reply code/core/write_sdl_ntuple.cc]: yes, I deleted that. I'm not sure why it still shows up in the conversation, maybe it's outdated.

**Review by VourMa — APPROVED** [review]
Thank you for the update, @YonsiG! Approving and merging.

**Pull request merged**